numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
```

Comment:
The iterator uses same kind casting. We do that for indices quite a bit unfortunatly, but.... Basically a 64bit int on a 32bit machine should work, and yes, it might have an integer overflow on casting, but we just hope nobody notices....